pkg/server/delta/v3/server.go

```diff
@@ -162,25 +162,25 @@ func (s *server) processDelta(str stream.DeltaStream, reqCh <-chan *discovery.De
 				// Since there was no previous state, we know we're handling the first request of this type
 				// so we set the initial resource versions if we have any, and also signal if this stream is in wildcard mode.
 				watch.state = stream.NewStreamState(len(req.GetResourceNamesSubscribe()) == 0, req.GetInitialResourceVersions())
-				watch.responses = make(chan cache.DeltaResponse, 1)
-
+			} else {
 				// We must signal goroutine termination to prevent a race between the cancel closing the watch
 				// and the producer closing the watch.
 				watch.Cancel()
+			}
 
-				s.subscribe(req.GetResourceNamesSubscribe(), watch.state.ResourceVersions)
-				s.unsubscribe(req.GetResourceNamesUnsubscribe(), watch.state.ResourceVersions)
+			s.subscribe(req.GetResourceNamesSubscribe(), watch.state.ResourceVersions)
+			s.unsubscribe(req.GetResourceNamesUnsubscribe(), watch.state.ResourceVersions)
 
-				watch.cancel = s.cache.CreateDeltaWatch(req, watch.state, watch.responses)
-				watches.deltaWatches[typeURL] = watch
+			watch.responses = make(chan cache.DeltaResponse, 1)
```

Comment:
One thining I am not completely certain is whether we should set `watch.responses` here or in the if above (after we set `watch.state`)

Reply:
I think this only matters in the case when in a single stream we receive a request for a particular type while a watch for the same type is open. My understanding is that this should never happen because envoy always waits for the response to the previous request before sending another request. 

However, if I am wrong, we should probably fix the following scenario:
1. The control plane receives a request for some typeUrl T, there are no open watchers for this type, so we create new state and open a watcher [here](https://github.com/s-matyukevich/go-control-plane/blob/fix-delta-watchers/pkg/server/delta/v3/server.go#L175)
2. The control plane doesn't respond immediately.
3. The control plane receives one more request for the same type T.
4. We cancel the existing watch [here](https://github.com/s-matyukevich/go-control-plane/blob/fix-delta-watchers/pkg/server/delta/v3/server.go#L168) fun [this](https://github.com/s-matyukevich/go-control-plane/blob/fix-delta-watchers/pkg/server/delta/v3/server.go#L178-L183) goroutine is still running because, as far as I see from [this](https://github.com/s-matyukevich/go-control-plane/blob/fix-delta-watchers/pkg/cache/v3/simple.go#L447-L453) code, canceling a delta watch doesn't result in anything being written to the `watch.responses` channel.
5. We are left with an orphaned goroutine that is going to run forever.

Reply:
Well delta might function differently, I think it's no longer a 1 -> 1 with the request cycle anymore. I think one of the bugs we were fixing was canceling the watch outside of the if. That will fix the case you have but we do need a way to clean up the go routine.

Reply:
@alecholmez  I think the right fix would be to close the responses channel [here](https://github.com/envoyproxy/go-control-plane/blob/main/pkg/cache/v3/simple.go#L447) and make sure we do the same in the linear cache implementation. This should clean up the goroutine. Do you agree? Should I add this additional fix to the existing PR?

Reply:
Well the responses channel is now propagated from the server since we changed the API for the cache. Should we just close it [here](https://github.com/envoyproxy/go-control-plane/blob/main/pkg/server/delta/v3/watches.go#L27) instead?

Reply:
Yep, that's even better. I'll update the code.

Reply:
Oops sorry I meant here: https://github.com/envoyproxy/go-control-plane/blob/main/pkg/server/delta/v3/watches.go#L45

@s-matyukevich

Reply:
ok, done

Reply:
Great thanks!